File: day03/day03-A.js
#!/usr/bin/env gjs

// Imports

imports.searchPath.unshift('../common');
const IO = imports.io;

// Functions

function count_columns(report) {
    let counts = new Array(report[0].length).fill(0);
    report.forEach(number => {
    	for (let index in number) {
    	    counts[index] += parseInt(number[index]);
	}
    });
    return counts;
}

function most_common(counts, total) {
    return counts.map(count => 2*count >= total ? '1' : '0')
}

function least_common(counts, total) {
    return counts.map(count => 2*count <  total ? '1' : '0')
}

// Main Execution

let report  = IO.readlines();
let counts  = count_columns(report)
let gamma   = parseInt(most_common(counts, report.length).join(''), 2);
let epsilon = parseInt(least_common(counts, report.length).join(''), 2);
IO.print(gamma * epsilon);

File: day03/day03-A.py
#!/usr/bin/env python3

import sys

# Functions

def count_columns(report):
    counts = [0]*len(report[0])
    for number in report:
        for index, digit in enumerate(number):
            counts[index] += int(digit)
    return counts

# Main Execution

def main():
    report  = [line.strip() for line in sys.stdin]
    counts  = count_columns(report)
    gamma   = int(''.join(['1' if count > len(report)//2 else '0' for count in counts]), 2)
    epsilon = int(''.join(['1' if count < len(report)//2 else '0' for count in counts]), 2)
    print(gamma * epsilon)

if __name__ == '__main__':
    main()
